### PR TITLE
Identify Prisma Version

### PR DESCRIPTION
**Pull request by do4gr**
The introspection command now returns a guesstimate of the db origin: 
 NonPrisma,
 Prisma1,
 Prisma11,
 Prisma2,

### CODE REVIEW

File: introspection-engine/connectors/sql-introspection-connector/src/calculate_datamodel.rs [hunk 1]
```diff
@@ -1,28 +1,80 @@
...
+    let migration_table = schema.tables.iter().any(|table| is_migration_table(&table));
+    let has_prisma_1_join_table = schema.tables.iter().any(|table| is_prisma_1_point_0_join_table(&table));
+    let has_prisma_1_1_or_2_join_table = schema
+        .tables
+        .iter()
+        .any(|table| is_prisma_1_point_1_or_2_join_table(&table));
+    let mut uses_on_delete = false;
+    let mut always_has_created_at_updated_at = true;
+    let mut uses_non_prisma_types = false;
+    let mut has_inline_relations = false;
```
Comment (tomhoule) [inline introspection-engine/connectors/sql-introspection-connector/src/calculate_datamodel.rs:26]: The state space of this function is getting pretty big :p

But it's hard to think of a good way to structure this without doing too much repetitive works (lots of looping) or introducing as much or more state.

File: introspection-engine/connectors/sql-introspection-connector/tests/db_specific_introspection/identify_version/mod.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            });
+            migration.create_table("Book", |t| {
+                t.add_column("id", types::primary());
+            });
+        })
+        .await;
+
+    let result = dbg!(api.introspect_version().await);
+    assert_eq!(result, Version::Prisma2);
+}
```
[the remark this answers is not in the document]
Comment (do4gr) [reply introspection-engine/connectors/sql-introspection-connector/tests/db_specific_introspection/identify_version/mod.rs:231]: No, there we used a separate db

File: introspection-engine/connectors/sql-introspection-connector/src/misc_helpers.rs [hunk 3]
```diff
@@ -13,7 +13,20 @@ pub fn is_migration_table(table: &Table) -> bool {
     table.name == "_Migration"
 }
 
-pub(crate) fn is_prisma_1_point_1_join_table(table: &Table) -> bool {
+pub(crate) fn is_relay_table(table: &Table) -> bool {
+    table.name == "_RelayId"
+        && table.columns[0].name == "id"
+        && table.columns[1].name.to_lowercase() == "stablemodelidentifier"
+}
```
[the remark this answers is not in the document]
Comment (do4gr) [reply introspection-engine/connectors/sql-introspection-connector/src/misc_helpers.rs:20]: Yeah, I could use two contains here.

File: introspection-engine/connectors/sql-introspection-connector/src/calculate_datamodel.rs (hunk 1 above)
Comment (do4gr) [reply introspection-engine/connectors/sql-introspection-connector/src/calculate_datamodel.rs:26]: Yeah, I also think it is pretty ugly but could not think of a good way to pull the whole functionality in a separate place without handing around references everywhere.

Comment (do4gr) [reply introspection-engine/connectors/sql-introspection-connector/src/calculate_datamodel.rs]: Reworked it after our discussion

**Pull request merged**